lib/engine/game/base.rb

```diff
@@ -2807,37 +2825,55 @@ def game_end_check_values
         self.class::GAME_END_CHECK
       end
 
-      def custom_end_game_reached?
-        false
+      def game_end_timing(reason)
+        self.class::GAME_END_CHECK[reason]
       end
 
       def game_end_check
-        triggers = {
-          all_closed: all_closed?,
-          bankrupt: bankruptcy_limit_reached?,
-          bank: @bank.broken?,
-          stock_market: @stock_market.max_reached?,
-          final_train: @depot.empty?,
-          final_phase: @phase&.phases&.last == @phase&.current,
-          custom: custom_end_game_reached?,
-        }.select { |_, t| t }
-
-        %i[immediate current_round current_or full_or one_more_full_or_set].each do |after|
-          triggers.keys.each do |reason|
-            if game_end_check_values[reason] == after
-              @final_turn ||= @turn + 1 if after == :one_more_full_or_set
-              return [reason, after]
-            end
-          end
+        return @game_end_trigger if self.class::GAME_END_LOCK_FIRST_REASON && @game_end_trigger
+
+        triggers = game_end_check_values.each_with_object([]) do |(reason, _after), t|
```

Comment:
Is it not possible to use `filter_map` here?

Reply:
I refactored this some more, with validating against the full DB I discovered some more bugs. Specifically there was a problem in 1822 games if the bank runs out and a corp reaches the end of the stock market on the same action.